pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,177 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// metricsWithDynamicLabels encompasses metrics that can have optional labels.
+type metricsWithDynamicLabels interface {
+	ObserveMetricWithDynamicLabels(metav1.Object, log.FieldLogger, map[string]string, float64)
+	RegisterMetricWithDynamicLabels()
+}
+
+type dynamicLabels struct {
+	// fixedLabels defines the label keys that are always registered on the metric.
+	fixedLabels []string
+	// optionalLabels is generated dynamically based on HiveConfig.Spec.MetricsConfig.MetricsWithDynamicLabels, and
+	// represents a map of metric label keys to ClusterDeployment metadata.label keys from which values will be gleaned
+	// when the metric is observed.
+	optionalLabels map[string]string
+}
+
+func WithFixedLabels(labels []string) func(*dynamicLabels) {
+	return func(d *dynamicLabels) {
+		d.fixedLabels = labels
+	}
+}
+
+func WithOptionalLabels(labels map[string]string) func(*dynamicLabels) {
+	return func(d *dynamicLabels) {
+		d.optionalLabels = labels
+	}
+}
+
+// getLabelList combines fixed and optional labels and returns the list without duplicates.
+func (d *dynamicLabels) getLabelList() []string {
+	var finalList []string
+
+	// Ensure there isn't an overlap between fixed and optional labels
+	keys := make(map[string]bool)
+	for _, entry := range append(d.fixedLabels, getKeys(d.optionalLabels)...) {
+		if _, value := keys[entry]; !value {
+			keys[entry] = true
+			finalList = append(finalList, entry)
+		}
+	}
+	return finalList
+}
+
+// buildLabels fetches the optional label values and merges the maps of fixedLabels and optionalLabels without
+// duplicates. The resultant map can be used for observing the metric.
+func (d *dynamicLabels) buildLabels(fixedLabels map[string]string, obj metav1.Object) prometheus.Labels {
+	optionalLabels := make(map[string]string, len(d.optionalLabels))
+	for _, v := range getKeys(d.optionalLabels) {
+		optionalLabels[v] = GetLabelValue(obj, d.optionalLabels[v])
+	}
+	labels := make(map[string]string, len(d.getLabelList()))
+	for label, value := range fixedLabels {
+		labels[label] = value
+	}
```

Comment:
As noted below, we could get rid of `defaultMissingLabels` if we changed this to:
```suggestion
	for _, label := range d.fixedLabels {
		if value := fixedLabels[label]; value != "" {
			labels[label] = value
		} else {
			labels[label] = constants.MetricLabelDefaultValue
		}
	}
```